crmsh/utils.py

```diff
@@ -2287,4 +2249,109 @@ def interface_choice():
     # should consider interface format like "ethx@xxx"
     interface_list = re.findall(r'(?:[0-9]+:) (.*?)(?=: |@.*?: )', out)
     return [nic for nic in interface_list if nic != "lo"]
+
+
+class InterfacesInfo(object):
+    def __init__(self, ipv6=False, second_heartbeat=False, specify_nic_list=[]):
+        self.ipv6 = ipv6
+        self._nic_info_dict = {}
+        self.second_heartbeat = second_heartbeat
+        self.get_interface_info()
+        self.specify_nic_list = specify_nic_list or self.get_default_nic_list_from_route()
+
+    def get_interface_info(self):
+        cmd = "ip -{} -o addr show".format(6 if self.ipv6 else 4)
+        rc, out, err = get_stdout_stderr(cmd)
+        if rc != 0:
+            raise ValueError(err)
+
+        # format on each line will like:
+        # 2: enp1s0    inet 192.168.122.241/24 brd 192.168.122.255 scope global enp1s0\       valid_lft forever preferred_lft forever
+        for line in out.splitlines():
+            _, nic, _, ip_with_mask, *_ = line.split()
+            if nic == "lo":
+                continue
+            # one nic might configured multi IP addresses
+            if nic not in self._nic_info_dict:
+                self._nic_info_dict[nic] = []
+            self._nic_info_dict[nic].append(Interface(ip_with_mask))
+
+        if not self._nic_info_dict:
+            raise ValueError("No address configured")
+        if self.second_heartbeat and len(self._nic_info_dict) == 1:
+            raise ValueError("Cannot configure second heartbeat line, since only one address is available")
+
+    def get_default_ip_list(self):
+        _ip_list = []
+        for nic in self.specify_nic_list:
+            if nic not in self.nic_list:
+                raise ValueError("Failed to detect IP address for {}".format(nic))
+            _ip_list.append(self._nic_info_dict[nic][0].ip)
+        if self.second_heartbeat and len(self.specify_nic_list) == 1:
+            for nic in self.nic_list:
+                if nic not in self.specify_nic_list:
+                    _ip_list.append(self._nic_info_dict[nic][0].ip)
+                    break
+        return _ip_list
+
+    def get_default_nic_list_from_route(self):
+        cmd = "ip -o route show"
+        rc, out, err = get_stdout_stderr(cmd)
+        if rc != 0:
+            raise ValueError(err)
+        res = re.search(r'^default via .* dev (.*?) ', out)
+        if res:
+            return [res.group(1)]
+        else:
+            common_warn("No default route configured")
```

Comment:
Maybe better `No default route configured. Using the first found nic`

Reply:
changed